analyzers/src/SonarAnalyzer.CSharp/Rules/TooManyLoggingCalls.cs

```diff
@@ -0,0 +1,217 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2024 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using SonarAnalyzer.Common.Walkers;
+
+namespace SonarAnalyzer.Rules.CSharp;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public sealed class TooManyLoggingCalls : ParametrizedDiagnosticAnalyzer
+{
+    private const string DiagnosticId = "S6664";
+    private const string MessageFormat = "Reduce the number of {0} logging calls within this code block from {1} to the {2} allowed.";
+
+    private static class CategoryNames
+    {
+        public const string Debug = "Debug";
+        public const string Error = "Error";
+        public const string Information = "Information";
+        public const string Warning = "Warning";
+    }
+
+    private static class DefaultTresholds
+    {
+        public const int Debug = 4;
+        public const int Error = 1;
+        public const int Information = 2;
+        public const int Warning = 1;
+    }
+
+    public static readonly DiagnosticDescriptor Rule = DescriptorFactory.Create(DiagnosticId, MessageFormat, isEnabledByDefault: false);
+
+    private static readonly ImmutableArray<KnownType> LoggerTypes = ImmutableArray.Create(
+        KnownType.Castle_Core_Logging_ILogger,
+        KnownType.log4net_ILog,
+        KnownType.log4net_Util_ILogExtensions,
+        KnownType.Microsoft_Extensions_Logging_ILogger,
+        KnownType.Microsoft_Extensions_Logging_LoggerExtensions,
+        KnownType.NLog_ILogger,
+        KnownType.NLog_ILoggerBase,
+        KnownType.NLog_ILoggerExtensions,
+        KnownType.Serilog_ILogger,
+        KnownType.Serilog_Log);
```

Comment:
Very small optimization:
Order these by popularity, to exit earlier at the type checks:

```csharp
        KnownType.Microsoft_Extensions_Logging_ILogger,
        KnownType.Microsoft_Extensions_Logging_LoggerExtensions,
        KnownType.NLog_ILogger,
        KnownType.NLog_ILoggerBase,
        KnownType.NLog_ILoggerExtensions,
        KnownType.Serilog_ILogger,
        KnownType.Serilog_Log,
        KnownType.log4net_ILog,
        KnownType.log4net_Util_ILogExtensions,
        KnownType.Castle_Core_Logging_ILogger
```